### ISSUE TITLE
OpenTSDB query failing

### ISSUE THREAD

**Issue by biswaKL** [OpenTSDB/opentsdb#1468]
{"error":{"code":424,"message":"10000 RPCs waiting

I am getting the above error and some times its taking lot of time to fetch the response

**Comment by biswaKL** [comment]
`{"error":{"code":424,"message":"Call queue is full on /10.247.72.145:60020, too many items queued`

**Comment by johnseekins** [comment]
You've overloaded HBase. If the call queue is full, HBase can't do any more processing.

**Comment by biswaKL** [comment]
now i have clustered my hbase with 2 nodes to distribute the load but i am getting new error
```
"code": 424,
"message": "[id: 0x82e2f455, /10.247.72.145:51740 :> /10.247.72.144:60020] got disconnected",

```

**Comment by johnseekins** [comment]
I'd wonder what HBase logs are saying. The log samples you are showing look like OpenTSDB logs.

That still looks like HBase not keeping up. What kind of scale are we talking about here? How much data are you writing?

**Comment by biswaKL** [comment]
Currently i am writing 50 to 60 records (values) per second.
I have 100+ threads sitting idle in both the node..!!

**Comment by johnseekins** [comment]
Well...the errors still _seem_ like HBase errors. What do the region server logs say? Also, is there any context around the errors you are getting? It's possible something else might actually be causing this that we're not seeing in the small amount of log you're showing.

**Comment by biswaKL** [comment]
in hbase while tailing the log for same request, i found this:

`2019-01-05 06:29:53,010 WARN  [RpcServer.FifoWFPBQ.default.handler=10,queue=10,port=60020] ipc.RpcServer: RpcServer.FifoWFPBQ.default.handler=10,queue=10,port=60020: caught a ClosedChannelException, this means that the server /10.247.72.144:60020 was processing a request but the client went away. The error message was: null
`

**Comment by johnseekins** [comment]
Well...those two logs together seem to imply the TSD process itself is dying. Where are you running the TSD process, how many TSDs, what does the TSD log look like, and what does your config look like?


The big question is: is there anything obvious pointing towards the TSD process dying or somerhing....

**Comment by biswaKL** [comment]
This is the log i found in opentsdb.log for the above query

```
08:17:10.655 INFO  [QueryStats.<init>] - Executing new query={"query":{"start":"1483312066","end":"1546297666","timezone":null,"options":null,"padding":false,"queries":[{"aggregator":"last","metric":"AQI.Aqi","tsuids":null,"downsample":"15m-avg","rate":false,"filters":[{"tagk":"Station_id","filter":"site_116","group_by":true,"type":"literal_or"}],"index":0,"rateOptions":null,"filterTagKs":[],"explicitTags":false,"tags":{"Station_id":"literal_or(site_116)"}}],"delete":false,"globalAnnotations":false,"noAnnotations":false,"showTSUIDs":false,"msResolution":false,"showQuery":false,"showStats":false,"showSummary":false,"useCalendar":false},"exception":"null","executed":1,"user":null,"requestHeaders":{"accept-language":"en-US,en;q=0.9","cookie":"*******","X-Forwarded-Host":"debug.cpcbccr.com:443","Connection":"close","upgrade-insecure-requests":"1","X-Forwarded-For":"157.41.192.142","Host":"https://upload.cpcbccr.com","cache-control":"max-age=0","accept-encoding":"gzip, deflate, br","user-agent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/69.0.3497.100 Safari/537.36","accept":"text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8"},"numRunningQueries":2,"httpResponse":null,"queryStartTimestamp":1546676230655,"queryCompletedTimestamp":0,"sentToClient":false,"stats":{}}
08:17:11.790 WARN  [HttpQuery.logWarn] - [id: 0x81ba4026, /157.41.192.142 => /10.247.72.145:4242] Bad Request on /api/query?start=1483312066&end=1546297666&m=last:15m-avg:AQI.Aqi{Station_id=site_116}: [id: 0xb8b349aa, /10.247.72.145:53984 :> /10.247.72.144:60020] got disconnected
08:17:11.792 INFO  [HttpQuery.logInfo] - [id: 0x81ba4026, /157.41.192.142 => /10.247.72.145:4242] HTTP /api/query?start=1483312066&end=1546297666&m=last:15m-avg:AQI.Aqi{Station_id=site_116} done in 1137ms
08:17:11.792 INFO  [QueryStats.markSent] - Completing query={"query":{"start":"1483312066","end":"1546297666","timezone":null,"options":null,"padding":false,"queries":[{"aggregator":"last","metric":"AQI.Aqi","tsuids":null,"downsample":"15m-avg","rate":false,"filters":[{"tagk":"Station_id","filter":"site_116","group_by":true,"type":"literal_or"}],"index":0,"rateOptions":null,"filterTagKs":["AAAM"],"explicitTags":false,"tags":{"Station_id":"literal_or(site_116)"}}],"delete":false,"globalAnnotations":false,"noAnnotations":false,"showTSUIDs":false,"msResolution":false,"showQuery":false,"showStats":false,"showSummary":false,"useCalendar":false},"exception":"[id: 0xb8b349aa, /10.247.72.145:53984 :> /10.247.72.144:60020] got disconnected\norg.hbase.async.RegionClient.cleanup(RegionClient.java:1172)","executed":1,"user":null,"requestHeaders":{"accept-language":"en-US,en;q=0.9","cookie":"*******","X-Forwarded-Host":"debug.cpcbccr.com:443","Connection":"close","upgrade-insecure-requests":"1","X-Forwarded-For":"157.41.192.142","Host":"https://upload.cpcbccr.com","cache-control":"max-age=0","accept-encoding":"gzip, deflate, br","user-agent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/69.0.3497.100 Safari/537.36","accept":"text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8"},"numRunningQueries":1,"httpResponse":{"code":424,"reasonPhrase":"Failed Dependency"},"queryStartTimestamp":1546676230655,"queryCompletedTimestamp":1546676231790,"sentToClient":true,"stats":{"avgHBaseTime":500.118663,"avgScannerTime":1132.706769,"avgScannerUidToStringTime":0.0,"dpsPostFilter":1152,"dpsPreFilter":1152,"maxHBaseTime":500.118663,"maxScannerUidtoStringTime":0.0,"processingPreWriteTime":1135.208135,"rowsPostFilter":1152,"rowsPreFilter":1152,"successfulScan":0,"totalTime":1137.363667,"uidPairsResolved":0}}

```

**Comment by johnseekins** [comment]
You have a 30 second timeout for _all_ queries set up in your config. That seems...aggressive. _And_ likely the cause of your disconnects.

**Comment by biswaKL** [comment]
Nope, i added that later, it was  failing before that as well, i tried by removing that, i get the same issue..

**Comment by biswaKL** [comment]
The problem got resolved when i upgraded from 2.3 to 2.4
Thanks guys for the support.

**Comment by bharathkuppala** [comment]
Hello, But still i see this issue in 2.4. Did you change any configuration?
